dali/operators/image/distortion/jpeg_compression_distortion_op_gpu.cu

```diff
@@ -38,17 +37,63 @@ class JpegCompressionDistortionGPU : public JpegCompressionDistortion<GPUBackend
   std::vector<int> quality_;
 };
 
+template <int ndim>
+TensorListShape<ndim - 1> unfold_dim(const TensorListShape<ndim> &shapes) {
+  static_assert(ndim > 1,
+                "Can't reduce dimentionality of dynamic, or single dimentional TensorListShape");
+
+  constexpr static int out_dim = ndim - 1;
+
+  using OutShapeType = TensorShape<out_dim>;
+  TensorListShape<out_dim> result;
+
+  for (int i = 0; i < shapes.size(); ++i) {
+    auto shape = shapes[i];
+    auto nfirst_dim = shape[0];
+    auto subshape = shape.last(out_dim);
+    std::vector<OutShapeType> unfolded_dim(nfirst_dim, subshape.template to_static<out_dim>());
+    result.append(unfolded_dim);
+  }
+
+  return result;
+}
+
+template <typename Type>
+TensorListView<StorageGPU, Type, 3> split_sequence(
+    const TensorListView<StorageGPU, Type, 4> &view) {
+  TensorListShape<3> new_shape = unfold_dim(view.shape);
+  return reinterpret<Type>(view, new_shape);
+  ;
+}
+
 void JpegCompressionDistortionGPU::RunImpl(workspace_t<GPUBackend> &ws) {
   const auto &input = ws.Input<GPUBackend>(0);
   auto &output = ws.Output<GPUBackend>(0);
-  auto in_view = view<const uint8_t, 3>(input);
-  auto out_view = view<uint8_t, 3>(output);
-  int nsamples = in_view.shape.size();
 
-  quality_.resize(nsamples);
-  for (int i = 0; i < nsamples; i++) {
-    quality_[i] = quality_arg_[i].data[0];
+  auto layout = input.GetLayout();
+  int nsamples = input.num_samples();
+
+  // Set quality argument for an image from samples
+  if (layout == "FHWC") {
+    quality_.resize(0);
+    for (int i = 0; i < nsamples; i++) {
+      auto nframes = input.tensor_shape_span(i)[0];
+      for (int j = 0; j < nframes; ++j) {
+        quality_.push_back(quality_arg_[i].data[0]);
+      }
+    }
+  } else {
+    quality_.resize(nsamples);
+    for (int i = 0; i < nsamples; i++) {
+      quality_[i] = quality_arg_[i].data[0];
+    }
   }
+
+  auto in_view = (layout == "FHWC") ? split_sequence(view<const uint8_t, 4>(input)) :
+                                      view<const uint8_t, 3>(input);
+  auto out_view =
+      (layout == "FHWC") ? split_sequence(view<uint8_t, 4>(output)) : view<uint8_t, 3>(output);
```

Comment:
I'd move this to the if/else above. For two reasons:
- It'd read better.
- You can split the sequence first, so that you know the number of samples in the output, and resize the `quality_` vector accordingly.
```
TensorListView<const uint8_t, 3> in_view;
TensorListView<uint8_t, 3> out_view;
if (layout.size() == 4) {
  in_view = split_sequence(view<const uint8_t, 4>(input));
  out_view = split_sequence(view<uint8_t, 4>(output));
  quality_.resize(out_view.size());
  // ...
} else {
  in_view = view<const uint8_t, 3>(input);
  out_view = view<uint8_t, 3>(output);
  quality_.resize(out_view.size());
  // ...
}
```

Reply:
Done